xls/delay_model/delay_model.proto

```diff
@@ -109,22 +117,77 @@ message DelayFactor {
   int64 operand_number = 2;
 }
 
+// Describes a nested mathematical expression combining delay factors (and possibly
+// constants). Useful when data for an op cannot be described by a linear
+// combination of individual delay factors.
+message DelayExpression {
+  // The expression is binary op combining
+  // two subexpressions.
+  enum BinaryOperation {
+    ADD = 0;
+    DIVIDE = 1;
+    MAX = 2;
+    MIN = 3;
+    MULTIPLY = 4;
+    POWER = 5;
+    SUB = 6;
+  }
+
+  // An expression can be either a BinaryOperation,
+  // a DelayFactor, or a int.
+  oneof expression_type {
+    BinaryOperation bin_op = 1;
+
+    // The expression is simply a delay factor.
+    DelayFactor factor = 2;
+
+    // The expression is a constant.
+    int64 constant = 3;
+  }
+
+  // Left and right hand side expressions combined by the BinaryOperation,
+  // if present.
+  DelayExpression lhs_expression = 4;
+  DelayExpression rhs_expression = 5;
+}
+
 // Describes a delay estimation model which fits a curve to the measured data
 // points for the operation. The function (curve) includes a constant term and
-// terms which are linear and logarithm functions of the indicated factors. For
-// example, if the factors are the result bit count and the operand count then
+// terms which are linear and logarithm functions of the indicated expressions. For
+// example, if the expressoins are the result bit count and the operand count then
 // the function for the estimated delay has the form:
 //
 //   delay_est = P_0 + P_1 * result_bit_count + P_2 * log2(result_bit_count) +
 //                     P_3 * operand_count    + P_4 * log2(operand_count)
 //
 // Where P_i are learned parameters.
 message RegressionEstimator {
-  repeated DelayFactor factors = 1;
+  repeated DelayExpression expressions = 1;
+
+  // K-fold cross-validation settings.
+  // The model will fail to build if it does not pass cross validation.
+  // Note: oneof used here because proto3 does not support optional
+  // fields without experimental features enabled.
+  oneof optional_num_cross_validation_folds {
```

Comment:
Done.  However, I'm not sure this is much cleaner.  This collects all kfold stuff into one message, but all fields within the kfold message will still be optional.  This is because 1) I'd like there to be a default number of folds, so specifying this is optional and 2) It would be reasonable to specify a max geomean error only, a max data mean error only, or specifiy maximums for both measures.  So, both fields should be optional since either could be ommitted / ignored.

Reply:
Do you really need default values? Can't the fields always be filled in, and if you get some nonsensical zero value because it is not filled in (like zero for number of cross validation folds) just raise an error?

Reply:
We don't strictly *need* a default value for the number of folds.  The goal there is just convenience.

However, I think max_data_point_error and max_fold_geomean_error need to be optional, since it is reasonable to not enforce one of these while enforcing the other (for my model, I endeed up enforcing a geomean error but not a data point error).